File: 20-valid-parentheses/20-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> st = new Stack<Character>();
        int n = s.length();
        if(n==1)
            return false;
        for(int i=0; i<n; i++){
            if(s.charAt(i) == '(' || s.charAt(i) == '{' || s.charAt(i) == '['){
                st.push(s.charAt(i));
            }
            if(s.charAt(i) == ')' ){
                if(st.size() == 0)
                    return false;
                if(st.peek() == '(')
                    st.pop();
                else
                    return false;
            }
            if(s.charAt(i) == '}' ){
                if(st.size() == 0)
                    return false;
                if(st.peek() == '{')
                    st.pop();
                else
                    return false;
            }
            if(s.charAt(i) == ']' ){
                if(st.size() == 0)
                    return false;
                if(st.peek() == '[')
                    st.pop();
                else
                    return false;
            }
        }
        if(st.size() != 0)
            return false;
        return true;
    }
}

File: 20-valid-parentheses/20-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        int n = s.length();
        int i=0;
        while(i<n){
            if(s[i] == '(' || s[i] == '[' || s[i] == '{')
                st.push(s[i]);
            else if(st.empty()){
                if(s[i] == ')' || s[i] == ']' || s[i] == '}')
                    return false;
            }else{
                if(st.top() == '(' && s[i] == ')')
                    st.pop();
                else if(st.top() == '[' && s[i] == ']')
                    st.pop();
                else if(st.top() == '{' && s[i] == '}')
                    st.pop();
                else
                    return false;
            }
            i++;
        }
        if(!st.empty())
            return false;
        return true;
    }
};